sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/LengthValidatingInputStream.java

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.core.implementation.http.rest;
+
+import com.azure.core.exception.UnexpectedLengthException;
+import com.azure.core.util.logging.ClientLogger;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Objects;
+
+import static com.azure.core.implementation.http.rest.RestProxyUtils.BODY_TOO_LARGE;
+import static com.azure.core.implementation.http.rest.RestProxyUtils.BODY_TOO_SMALL;
+
+/**
+ * An {@link InputStream} decorator that tracks the number of bytes read from an inner {@link InputStream} and throws
+ * an exception if the number of bytes read doesn't match what was expected.
+ */
+final class LengthValidatingInputStream extends InputStream {
+    private final InputStream inner;
+    private final long expectedReadSize;
+
+    private long position;
+    private long mark = -1;
+
+    /**
+     * Creates a new {@link LengthValidatingInputStream}.
+     *
+     * @param inputStream The {@link InputStream} being decorated.
+     * @param expectedReadSize The expected number of bytes to be read from the inner {@code inputStream}.
+     */
+    LengthValidatingInputStream(InputStream inputStream, long expectedReadSize) {
+        this.inner = Objects.requireNonNull(inputStream, "'inputStream' cannot be null.");
+
+        if (expectedReadSize < 0) {
+            throw new ClientLogger(LengthValidatingInputStream.class)
+                .logExceptionAsError(new IllegalArgumentException("'expectedReadSize' cannot be less than 0."));
+        }
+
+        this.expectedReadSize = expectedReadSize;
+    }
+
+    @Override
+    public synchronized int read(byte[] b, int off, int len) throws IOException {
+        int readSize = inner.read(b, off, len);
+        validateLength(readSize);
+
+        return readSize;
+    }
+
+    @Override
+    public synchronized long skip(long n) throws IOException {
+        long skipped = inner.skip(n);
+        position += skipped;
+        return skipped;
+    }
+
+    @Override
+    public int available() throws IOException {
+        return inner.available();
+    }
+
+    @Override
+    public void close() throws IOException {
+        inner.close();
+    }
+
+    @Override
+    public synchronized void mark(int readlimit) {
+        inner.mark(readlimit);
+        mark = position;
+    }
+
+    @Override
+    public synchronized void reset() throws IOException {
+        inner.reset();
+        position = mark;
+    }
+
+    @Override
+    public boolean markSupported() {
+        return inner.markSupported();
+    }
+
+    @Override
+    public synchronized int read() throws IOException {
+        int read = inner.read();
+        validateLength(read == -1 ? -1 : 1);
+
+        return read;
+    }
+
+    private void validateLength(int readSize) {
+        if (readSize == -1) {
```

Comment:
There is a corner case here when the consumer of this doesn't read the entire stream and instead reads based on the `Content-Length` of the request instead. Do we need to add any additional logic, such as reading one byte ahead of the data requested to make sure we don't let the following through:

- Create a request based on an InputStream with 4097 bytes
- Request has `Content-Length` 4096
- Networking layer has logic to only read the amount based on the `Content-Length`

Where validate length never fails the request as the 4097th byte is never read.

Reply:
all our clients hit eof when reading input streams. I added test for this. 
I think we shouldn't complicate logic here and introduce extra I/O when we hit content length if they anyway hit eof.